Move serial input handling to its own thread

bryteise opened the pull request:
Likely needs major changes beyond the revamp of the Serial struct but
initial layout of the change is here.

This fails due to the queue_input_bytes function being a method
implemented for Serial which can't be passed to the spawn as its
lifetime isn't static.

One option is to move most of the Serial struct implementation and
data to the SerialEpollHandler to allow for the queue_input_bytes
method to be run in SerialEpollHandler's handle_event.

I'm open to other suggestions of course and wider suggestions for
improvements as this is a lot of copying from various parts of the
virtio-devices portion of the code base that could likely be done
better.

rbradford commented:
@bryteise You need to think about something like a `SerialManager` that sits atop the current `Serial` or `Pl011`. All it does is spawn an even loop listening on the input source and inject into the real serial implementation. Trying to modify the existing `Serial` struct is the wrong direction as you would have to do the same for the `Pl011` version too.

rbradford commented:
That `SerialManager` can just live adjacent to `DeviceManager` in the vmm crate. I don't think you should be making any changes to the `devices` crate.

bryteise commented:
@rbradford I should add, I was originally going to propagate the `DeviceManager`'s `exit_evt` but didn't really get an idea of how that would be useful looking at the virtio console code.

rbradford commented on vmm/src/device_manager.rs:
```diff
@@ -1017,6 +1009,39 @@ impl DeviceManager {
+
+        let thread = thread::Builder::new()
+            .name("serial-manager".to_string())
+            .spawn(move || {
```
You want to use the `exit_evt` to write to when the thread panics (using `std::panic::catch_unwind`)

bryteise replied:
Ah okay, I am reusing the `DeviceManager`'s `exit_evt` as it didn't seem to make much sense to have a different one given the shutdown probably would still be the way to respond to the evt write.

rbradford commented on vmm/src/serial_manager.rs:
```diff
@@ -0,0 +1,235 @@
+        };
+        if serial_tty && unsafe { libc::isatty(libc::STDIN_FILENO as i32) } != 0 {
+            epoll.add_stdin().map_err(Error::Epoll)?;
+        }
```
You can't have the serial attached to PTY and TTY so you should a match on the mode from the `ConsoleConfig`

rbradford commented on vmm/src/serial_manager.rs:
```diff
@@ -0,0 +1,235 @@
+        Ok(EpollContext { epoll_file })
+    }
+
+    pub fn add_stdin(&mut self) -> result::Result<(), io::Error> {
```
I don't think you need a special case for stdin here it's got a FD number like any other file.

rbradford commented on vmm/src/serial_manager.rs:
```diff
@@ -0,0 +1,235 @@
+
+impl SerialManager {
+    #[cfg(target_arch = "x86_64")]
+    pub fn new(serial: Arc<Mutex<Serial>>, pty: Option<File>, serial_tty: bool) -> Result<Self> {
```
This also handles the case of the serial just going to a file.

rbradford commented on vmm/src/serial_manager.rs:
```diff
@@ -0,0 +1,235 @@
+    epoll_file: File,
+}
+
+impl EpollContext {
```
Don't think you really need this struct. Just store the epoll_file.

rbradford commented on vmm/src/serial_manager.rs:
```diff
@@ -0,0 +1,235 @@
+        Ok(())
+    }
+
+    fn handle_pty(&mut self) -> Result<()> {
```
This should have the same SAC handling as stdin. It's a bug that it doesn't.

rbradford commented on vmm/src/serial_manager.rs:
```diff
@@ -0,0 +1,235 @@
+        Ok(())
+    }
+
+    fn handle_stdin(&self) -> Result<()> {
```
You only need one function to read from the active `File` either PTY, TTY (stdin) or real file on disk.

rbradford commented on vmm/src/serial_manager.rs:
```diff
@@ -0,0 +1,235 @@
+                        let event = event.data;
+                        warn!("Unknown serial manager loop event: {}", event);
+                    }
+                    EpollDispatch::Stdin => {
```
You only need one event for when the file has something to read from.

bryteise commented:
This isn't quite right when handling input on the tty, for some reason it seems to drop bytes. However if I `debug!` the `handle_in` function whenever it gets called the bytes aren't dropped so I've got something dumb I'm doing. I tried a little test on reading from `io::stdin` and using a `lock().read_raw()` call chain but the same behavior was observed (thankfullyish).

Would appreciate if folks have any ideas about that.

rbradford commented:
> This isn't quite right when handling input on the tty, for some reason it seems to drop bytes. However if I `debug!` the `handle_in` function whenever it gets called the bytes aren't dropped so I've got something dumb I'm doing. I tried a little test on reading from `io::stdin` and using a `lock().read_raw()` call chain but the same behavior was observed (thankfullyish).
> 
> Would appreciate if folks have any ideas about that.

I couldn't reproduce your issue but I noticed you weren't opening in non-blocking mode which will affect the read call as it may block.

Also I noticed that input after a reboot with serial on tty doesn't work. This is usually because the file descriptor for stdin has been closed.

bryteise commented:
Oops I realized you can't just drop the `JoinHandle` and expect the thread to exit. I am killing the thread now and and recreating on reboot and am able to use input still.

rbradford commented on vmm/src/vm.rs:
```diff
@@ -1913,6 +1917,20 @@ impl Vm {
+                    self.device_manager
+                        .lock()
+                        .unwrap()
+                        .start_serial_manager_thread(mode)
```
Could this just be part of `add_console_device()` function in `DeviceManager`?

bryteise replied:
I wasn't exactly sure about the timing of when to start the thread. I'll try that out.

rbradford commented on vmm/src/serial_manager.rs:
```diff
@@ -0,0 +1,283 @@
+        }
+    }
+
+    #[cfg(target_arch = "aarch64")]
```
These constructors are basically identical. Can't you put the attribute on the function parameter instead?

rbradford replied:
(I pushed a change that did that.)

bryteise replied:
Thanks!

rbradford commented:
@bryteise I think this needs a rebase and the GH actions seem upset.

bryteise commented:
I rebased and squashed your changes as they stopped rebasing nicely. This works but I notice that sometimes the kill_evt sometimes will not be handled by the thread.

I had a bad activation from virtio console that looked suspiciously similar and when I looked at the rust log (with -v) it seemed as though the `KILL_EVENT` was handled well into the vm booting up:

```
cloud-hypervisor: 43.296603819s: <vcpu0> INFO:vmm/src/cpu.rs:796 -- VmExit::Reset
cloud-hypervisor: 43.296695086s: <vmm> INFO:vmm/src/lib.rs:1279 -- VM reset event
cloud-hypervisor: 43.296790419s: <vmm> INFO:virtio-devices/src/device.rs:356 -- Resuming virtio-console
cloud-hypervisor: 43.296834067s: <vmm> INFO:virtio-devices/src/device.rs:356 -- Resuming virtio-block
cloud-hypervisor: 43.296873326s: <vmm> INFO:virtio-devices/src/device.rs:356 -- Resuming virtio-rng
cloud-hypervisor: 43.297083315s: <_console> INFO:virtio-devices/src/epoll_helper.rs:134 -- KILL_EVENT received, stopping epoll loop
cloud-hypervisor: 43.297095005s: <_rng> INFO:virtio-devices/src/epoll_helper.rs:134 -- KILL_EVENT received, stopping epoll loop
cloud-hypervisor: 43.297105469s: <_disk0_q0> INFO:virtio-devices/src/epoll_helper.rs:134 -- KILL_EVENT received, stopping epoll loop
cloud-hypervisor: 43.373340708s: <vmm> INFO:vmm/src/memory_manager.rs:1257 -- Created userspace mapping: 0 -> 7f0b70000000 80000000
cloud-hypervisor: 43.373426592s: <vmm> INFO:vmm/src/vm.rs:516 -- Booting VM from config: Mutex { data: VmConfig { cpus: CpusConfig { boot_vcpus: 1, max_vcpus: 1, topology: None, kvm_hyperv: false, max_phys_bits: None }, memory: MemoryConfig { size: 2147483648, mergeable: false, hotplug_method: Acpi, hotplug_size: None, hotplugged_size: None, shared: false, huge
pages: false, hugepage_size: None, zones: None }, kernel: Some(KernelConfig { path: "/home/wdouglas/tmp/hypervisor-fw" }), initramfs: None, cmdline: CmdlineConfig { args: "" }, disks: Some([DiskConfig { path: Some("/home/wdouglas/tmp/cur.raw"), readonly: false, direct: false, iommu: false, num_queues: 1, queue_size: 128, vhost_user: false, vhost_socket: None, p
oll_queue: true, rate_limiter_config: None, id: Some("_disk0"), disable_io_uring: false }]), net: None, rng: RngConfig { src: "/dev/urandom", iommu: false }, balloon: None, fs: None, pmem: None, serial: ConsoleConfig { file: None, mode: Tty, iommu: false }, console: ConsoleConfig { file: None, mode: Null, iommu: false }, devices: None, user_devices: None, vsock
: None, iommu: false, sgx_epc: None, numa: None, watchdog: false }, poisoned: false, .. } 
cloud-hypervisor: 43.374596663s: <vmm> INFO:vmm/src/device_manager.rs:1908 -- Creating virtio-block device: DiskConfig { path: Some("/home/wdouglas/tmp/cur.raw"), readonly: false, direct: false, iommu: false, num_queues: 1, queue_size: 128, vhost_user: false, vhost_socket: None, poll_queue: true, rate_limiter_config: None, id: Some("_disk0"), disable_io_uring: 
false }                                                                                   
cloud-hypervisor: 43.420345345s: <vmm> INFO:vmm/src/device_manager.rs:1989 -- Using asynchronous RAW disk file (io_uring)
cloud-hypervisor: 43.420452029s: <vmm> INFO:vmm/src/device_manager.rs:2220 -- Creating virtio-rng device: RngConfig { src: "/dev/urandom", iommu: false }
cloud-hypervisor: 43.420762554s: <vmm> INFO:vmm/src/vm.rs:1860 -- Booting VM
cloud-hypervisor: 43.42080419s: <vmm> INFO:vmm/src/vm.rs:959 -- Loading kernel
cloud-hypervisor: 43.420923653s: <vmm> INFO:vmm/src/vm.rs:985 -- Kernel loaded: entry_addr = 0x11205e
cloud-hypervisor: 43.420960078s: <vmm> INFO:vmm/src/cpu.rs:681 -- Request to create new vCPUs: desired = 1, max = 1, allocated = 0, present = 0
cloud-hypervisor: 43.42100665s: <vmm> INFO:vmm/src/cpu.rs:639 -- Creating vCPU: cpu_id = 0 
cloud-hypervisor: 43.421126609s: <vmm> INFO:vmm/src/cpu.rs:276 -- Configuring vCPU: cpu_id = 0
cloud-hypervisor: 43.421201839s: <vmm> INFO:vmm/src/vm.rs:994 -- Configuring system
cloud-hypervisor: 43.421672082s: <vmm> INFO:vmm/src/vm.rs:1016 -- Created ACPI tables: rsdp_addr = 0xa0000
cloud-hypervisor: 43.4217231s: <vmm> INFO:vmm/src/cpu.rs:859 -- Starting vCPUs: desired = 1, allocated = 1, present = 0
cloud-hypervisor: 43.421768848s: <vmm> INFO:vmm/src/cpu.rs:720 -- Starting vCPU: cpu_id = 0
cloud-hypervisor: 43.431012414s: <vcpu0> INFO:virtio-devices/src/transport/pci_device.rs:1049 -- _virtio-pci-_disk0: Needs activation; writing to activate event fd
cloud-hypervisor: 43.431086758s: <vcpu0> INFO:virtio-devices/src/transport/pci_device.rs:1054 -- _virtio-pci-_disk0: Needs activation; returning barrier
cloud-hypervisor: 43.43113566s: <vcpu0> INFO:vmm/src/vm.rs:409 -- Waiting for barrier
cloud-hypervisor: 43.431193443s: <vmm> INFO:vmm/src/lib.rs:1287 -- Trying to activate pending virtio devices: count = 1
cloud-hypervisor: 43.431250963s: <vmm> INFO:virtio-devices/src/transport/pci_device.rs:697 -- _virtio-pci-_console: Device does not need activation
cloud-hypervisor: 43.431322403s: <vmm> INFO:virtio-devices/src/block.rs:484 -- Changing cache mode to writethrough
cloud-hypervisor: 43.440374331s: <vmm> INFO:virtio-devices/src/transport/pci_device.rs:693 -- _virtio-pci-_disk0: Waiting for barrier
cloud-hypervisor: 43.440450268s: <vmm> INFO:virtio-devices/src/transport/pci_device.rs:695 -- _virtio-pci-_disk0: Barrier released
cloud-hypervisor: 43.440499195s: <vmm> INFO:virtio-devices/src/transport/pci_device.rs:697 -- _virtio-pci-_rng: Device does not need activation
cloud-hypervisor: 43.440571293s: <vcpu0> INFO:vmm/src/vm.rs:411 -- Barrier released
cloud-hypervisor: 48.481774772s: <vcpu0> INFO:virtio-devices/src/device.rs:356 -- Resuming virtio-block
cloud-hypervisor: 48.481845556s: <_disk0_q0> INFO:virtio-devices/src/epoll_helper.rs:134 -- KILL_EVENT received, stopping epoll loop
```

Note the `KILL_EVENT` for `_disk0_q0` showing up twice, one well after the boot seems to be going. This happens consistently though so I'm not sure if that's a real problem or not. Either way happy to open an issue for this.

And this is the error from the VM's perspective:

```
PCI host bridge to bus 0000:00                                                                                                                                                                                                                                                                                                                                             
pci_bus 0000:00: root bus resource [bus 00-ff]                                                                                                                                                                                                                                                                                                                             
pci_bus 0000:00: root bus resource [io  0x0000-0x0cf7 window]                                                                                                                                                                                                                                                                                                              
pci_bus 0000:00: root bus resource [io  0x0d00-0xffff window]                                                                                                                                                                                                                                                                                                              
pci_bus 0000:00: root bus resource [mem 0xc0000000-0xe7ffffff window]                     
pci_bus 0000:00: root bus resource [mem 0x100000000-0x7ffffeffff window]                                                                                                             
pci 0000:00:00.0: [8086:0d57] type 00 class 0x060000                                                                                                                                 
pci 0000:00:01.0: [1af4:1043] type 00 class 0xffff00                                                                                                                                 
pci 0000:00:01.0: reg 0x10: [mem 0x7ffff00000-0x7ffff7ffff 64bit]                         
pci 0000:00:02.0: [1af4:1042] type 00 class 0x018000                                                                                                                                 
pci 0000:00:02.0: reg 0x10: [mem 0xe7f80000-0xe7ffffff]                                   
pci 0000:00:03.0: [1af4:1044] type 00 class 0xffff00                                      
pci 0000:00:03.0: reg 0x10: [mem 0x7fffe80000-0x7fffefffff 64bit]                                                                                                                    
SCSI subsystem initialized                                                                                                                                                           
vgaarb: loaded                                                                            
Registered efivars operations                                                                                                                                                        
NetLabel: Initializing                                                                                                                                                               
NetLabel:  domain hash size = 128                                                         
NetLabel:  protocols = UNLABELED CIPSOv4 CALIPSO                                          
NetLabel:  unlabeled traffic allowed by default                                                                                                                                      
PCI: Using ACPI for IRQ routing                                                                                                                                                      
clocksource: Switched to clocksource kvm-clock                                                                                                                                       
VFS: Disk quotas dquot_6.6.0                                                                                                                                                         
VFS: Dquot-cache hash table entries: 512 (order 0, 4096 bytes)                                                                                                                       
AppArmor: AppArmor Filesystem Enabled                                                                                                                                                
pnp: PnP ACPI init                                                                                                                                                                                                                                                                                                                                                         
system 00:00: [mem 0xe8000000-0xf7ffffff] has been reserved                                                                                                                          
pnp: PnP ACPI: found 2 devices                                                                                                                                                       
NET: Registered protocol family 2                                                                                                                                                    
IP idents hash table entries: 32768 (order: 6, 262144 bytes, linear)                      
tcp_listen_portaddr_hash hash table entries: 1024 (order: 2, 16384 bytes, linear)                                                                                                    
TCP established hash table entries: 16384 (order: 5, 131072 bytes, linear)                
TCP bind hash table entries: 16384 (order: 6, 262144 bytes, linear)                                                                                                                  
TCP: Hash tables configured (established 16384 bind 16384)                                
MPTCP token hash table entries: 2048 (order: 3, 49152 bytes, linear)                      
UDP hash table entries: 1024 (order: 3, 32768 bytes, linear)                                                                                                                         
UDP-Lite hash table entries: 1024 (order: 3, 32768 bytes, linear)                                                                                                                                                                                                                                                                                                          
NET: Registered protocol family 1                                                                                                                                                    
NET: Registered protocol family 44                                                                                                                                                   
pci_bus 0000:00: resource 4 [io  0x0000-0x0cf7 window]                                                                                                                               
pci_bus 0000:00: resource 5 [io  0x0d00-0xffff window]                                                                                                                               
pci_bus 0000:00: resource 6 [mem 0xc0000000-0xe7ffffff window]                                                                                                                       
pci_bus 0000:00: resource 7 [mem 0x100000000-0x7ffffeffff window]                                                                                                                    
PCI: CLS 0 bytes, default 64                                                                                                                                                         
RAPL PMU: API unit is 2^-32 Joules, 0 fixed counters, 10737418240 ms ovfl timer                                                                                                      
clocksource: tsc: mask: 0xffffffffffffffff max_cycles: 0x19f2297dd97, max_idle_ns: 440795236593 ns                                                                                                                                                                                                                                                                         
clocksource: Switched to clocksource tsc                                                                                                                                             
platform rtc_cmos: registered platform RTC device (no PNP device found)                                                                                                              
Initialise system trusted keyrings                                                                                                                                                   
Key type blacklist registered                                                                                                                                                        
workingset: timestamp_bits=40 max_order=19 bucket_order=0                                 
squashfs: version 4.0 (2009/01/31) Phillip Lougher                                                                                                                                   
9p: Installing v9fs 9p2000 file system support                                                                                                                                       
Key type asymmetric registered                                                                                                                                                       
Asymmetric key parser 'x509' registered                                                                                                                                              
Block layer SCSI generic (bsg) driver version 0.4 loaded (major 250)                                                                                                                 
io scheduler mq-deadline registered                                                                                                                                                                                                                                                                                                                                        
io scheduler kyber registered                                                                                                                                                                                                                                                                                                                                              
shpchp: Standard Hot Plug PCI Controller Driver version: 0.4                                                                                                                         
input: Power Button as /devices/LNXSYSTM:00/LNXSYBUS:00/PNP0C0C:00/input/input0                                                                                                      
ACPI: Power Button [PWRB]                                                                                                                                                            
virtio-pci 0000:00:01.0: enabling device (0000 -> 0002)                                                                                                                              
virtio-pci 0000:00:02.0: enabling device (0000 -> 0002)                                                                                                                                                                                                                                                                                                                    
virtio-pci 0000:00:03.0: enabling device (0000 -> 0002)                                                                                                                                                                                                                                                                                                                    
Serial: 8250/16550 driver, 4 ports, IRQ sharing disabled                                                                                                                                                                                                                                                                                                                   
00:01: ttyS0 at I/O 0x3f8 (irq = 24, base_baud = 115200) is a U6_16550A                                                                                                              
virtio_blk virtio1: [vda] 4612096 512-byte logical blocks (2.36 GB/2.20 GiB)                                                                                                                                                                                                                                                                                               
thread 'vmm' panicked at 'Failed to activate device: BadActivate', virtio-devices/src/transport/pci_device.rs:691:29                                                                 
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace             
Failed to join on VMM thread: Any { .. } 
```

rbradford commented:
> Note the `KILL_EVENT` for `_disk0_q0` showing up twice, one well after the boot seems to be going. This happens consistently though so I'm not sure if that's a real problem or not. Either way happy to open an issue for this.

This is because you're using a firmware. So the device is reset by the booted kernel. i.e. the virtio-block thread is created first for the activation made by the firmware. Later the loaded kernel then resets the device which kills that thread and then activates it again (since the virtio configuration is completely different for the two different users.)

rbradford commented on vmm/src/device_manager.rs:
```diff
@@ -4401,5 +4452,11 @@ impl Drop for DeviceManager {
         for (device, _, _) in self.virtio_devices.drain(..) {
             device.lock().unwrap().shutdown();
         }
+        if let Some(serial_kill_evt) = &self.serial_kill_evt {
```
`.take()`

rbradford commented on vmm/src/device_manager.rs:
```diff
@@ -4401,5 +4452,11 @@ impl Drop for DeviceManager {
+                "stopping serial-manager thread: {:?}",
+                serial_kill_evt.write(1)
+            );
+        }
```
Now `.join` on the thread handle to wait for the thread to finish.

bryteise replied:
Ah right, I completely forgot about this as I went through a few restructures X(.

rbradford reviewed (commented):
I think you should hide all the thread starting/shutdown inside the `SerialManager` struct which if you store on the `DeviceManager` will get dropped and then will result in fewer changes to DeviceManager.

bryteise commented:
Ah thanks for insight into the firmware usage and that makes total sense, I forget that the fw is poke the virtio disk X(. Still not sure what the crash I've seen was related to was.

I'm pretty happy with the latest commit, I greatly appreciate the feedback and moving the thread handling more into the SerialManager makes everything much nicer.

rbradford reviewed (approved):
Looks good, this change looks a lot more minimal (less intrusive) than before which is a good sign.

rbradford commented:
@sboeuf I'd like you to take a look as @bryteise and I have been looking at this for some time and may not see things that are non-obvious.

sboeuf commented:
@rbradford 
>@sboeuf I'd like you to take a look as @bryteise and I have been looking at this for some time and may not see things that are non-obvious.

Of course, let me look at it in details!

sboeuf commented on vmm/src/serial_manager.rs:
```diff
@@ -0,0 +1,245 @@
+// Copyright © 2021 Intel Corporation
```
nit: This file could have been added in a separate commit (prior to this one).

sboeuf commented on vmm/src/serial_manager.rs:
```diff
@@ -0,0 +1,245 @@
+                        .try_clone()
+                        .map_err(Error::FileClone)?
+                } else {
+                    return Ok(None);
```
Shouldn't this be treated as an error? I mean if the mode is Pty and we haven't been given an actual PtyPair, this should raise a flag.

rbradford replied:
I think after a restore the `PtyPair` is None.

bryteise replied:
@rbradford I thought `PtyPair` should still be set assuming the `serial_pty` was correctly passed though from the vmm?

@sboeuf It could be an error. I ended up being quite forgiving on creating a `SerialManager` given it is only handling input and it may be reasonable to have a pty device be removable in the future.

sboeuf replied:
Fair enough :)

sboeuf commented on vmm/src/serial_manager.rs:
```diff
@@ -0,0 +1,245 @@
+                    return Ok(None);
+                }
+            }
+            _ => return Ok(None),
```
Same here, the caller shouldn't try to create a new SerialManager when the mode is not appropriate, therefore we should throw an error here as well.

rbradford replied:
This is kinda non-obvious but the `SerialManager` is only created (at the moment) where the stdin is a tty. Imagine a situation where you run cloud-hypervisor with --serial tty but have already closed stdin (e.g. like `nohup`) and stdout is redirected to file.

rbradford replied:
The check on the `isatty()` or on the `PtyPair` could be done in the DeviceManager. Or maybe this `::new()` function just needs a comment.

bryteise replied:
@sboeuf For the line 123 return, similarly to the above, I could have made the return an error as the code calling this in `device_manager` checks specifically for the `mode` before calling. I decided to `return Ok(None)` though as that check could reasonably be removed and not creating a `SerialManager` instead of returning an `Error` seemed correct as a `Null` mode would have been a potentially valid case.

Just a note, the different checks needed given the combination of options accepted (`serial_pty` is `Some` and the `mode` is `Pty` `||` `mode` is `Tty` and stdin is as well) I felt were ugly putting into `device_manager`. `device_manager`'s calling of the `SerialManager::new` and `SerialManager::start_thread` are already a pretty ugly bit of code anyway =/. I mostly felt this was my least worst option after poking it a bit but I'm happy to make changes.

sboeuf replied:
Yeah I understand. The whole machine state is quite complex and if you feel that's the best we can do, let's go with that!

sboeuf commented on vmm/src/serial_manager.rs:
```diff
@@ -0,0 +1,245 @@
+
+    pub fn start_thread(&mut self, exit_evt: EventFd) -> Result<()> {
+        // Don't allow this to be run if the handle exists
+        if self.handle.is_some() {
+            return Ok(());
+        }
```
Nice, you're handling the case where this function would be called multiple times. Might be worth adding a warning message here because we don't expect this to happen.

sboeuf commented on vmm/src/serial_manager.rs:
```diff
@@ -0,0 +1,245 @@
+            .name("serial-manager".to_string())
+            .spawn(move || {
+                std::panic::catch_unwind(AssertUnwindSafe(move || {
+                    const EPOLL_EVENTS_LEN: usize = 100;
```
I know we have this through our codebase (I'm talking about the choice for using 100), but I'm not sure that's needed. Could you try to check if we could provide a more meaningful value?

rbradford replied:
Yeh, i spotted this too.

bryteise replied:
I made it 3 just because we do have a case for Unknown.

sboeuf reviewed (commented):
@bryteise @rbradford this looks like a great PR to me. Having a dedicated thread is way better IMO. Nothing wrong jumped to my eyes, just a few nits here and there.
